scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb

```diff
@@ -15,3 +15,59 @@ Sinks are last in the [pipeline][docs.pipelines], responsible for sending
 <%= components_table(metadata.sinks.to_h.values.sort) %>
 
 [+ request a new sink][url.new_sink]
+
+## Healthchecks
+
+Sinks may implement a healthcheck as a means for validating their configuration
+against the envionment and external systems. Ideally, this allows the system to
+inform users of problems such as insufficient credentials, unreachable
+endpoints, non-existant tables, etc. They're not perfect, however, since it's
+impossible to exhaustively check for issues that may happen at runtime.
+
+### Guidelines for writing healthchecks
+
+When implementing healthchecks, we prefer false positives to false negatives.
+This means we would prefer that a healthcheck pass and the sink then fail than
+to have the healthcheck fail when the sink would have been able to run
+successfully.
+
+A common cause of false negatives in healthchecks is performing an operation
+that the sink itself does not need. For example, listing all of the available S3
+buckets and checking that the configured bucket is in that list. The S3 sink
+doesn't need the ability to list all buckets, and a user that knows that may not
+have given it permission to do so. In that case, the healthcheck will fail due
+to bad credentials even through its credentials are sufficient for normal
+operation.
+
+This leads to a general strategy of mimicking what the sink itself does.
+Unfortunately, the fact that healthchecks don't have real events available to
+them leads to some limitations here. The most obvious example of this is with
+sinks where the exact target of a write depends on the value of some field in
+the event (e.g. an interpolated Kinesis stream name). It also pops up for sinks
+where incoming events are expected to conform to a specific schema. In both
+cases, random test data is reasonably likely to trigger a potentially false
+negative result. Even in simpler cases, we need to think about the effects of
+writing test data and whether the user would find that surprising or invasive.
+The answer usually depends on the system we're interfacing with.
+
+In some cases, like the Kinesis example above, the right thing to do might be
+nothing at all. If we require dynamic information to figure out what entity
+(i.e. Kinesis stream in this case) that we're even dealing with, odds are very
+low that we'll be able to come up with a way to meaningfully validate that it's
+in working order. It's perfectly valid to have a healthcheck that falls back to
```

Comment:
So I'm assuming this is kinda the main point, we are happy to do nothing at all if we can't reliably not cause false positives and data pollution? What worries me here is as we create more advanced sinks health checks will start to become almost irrelevant since there is a lot of dynamicness going on?

Reply:
That's possible, yeah. I think there will always be popular sinks for which they're useful, but if turn out not to be a big deal that's ok too. Unless we want to compromise on the false positive and data pollution values, I don't think we really have another option.